SYCL/KernelFusion/internalize_vfunc.cpp

```diff
@@ -0,0 +1,88 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// UNSUPPORTED: cuda || hip
+// REQUIRES: fusion
+
+// Test complete fusion with private internalization specified on the
+// accessors for a device kernel with sycl::vec::load and sycl::vec::store.
+
+#define VEC 4
+
+#include <sycl/sycl.hpp>
+
+using namespace sycl;
+
+int main() {
+  constexpr size_t numVec = 512;
+  constexpr size_t dataSize = numVec * VEC;
+  int in1[dataSize], in2[dataSize], in3[dataSize], tmp[dataSize], out[dataSize];
+
+  for (size_t i = 0; i < dataSize; ++i) {
+    in1[i] = i * 2;
+    in2[i] = i * 3;
+    in3[i] = i * 4;
+    tmp[i] = -1;
+    out[i] = -1;
+  }
+
+  queue q{ext::codeplay::experimental::property::queue::enable_fusion{}};
+
+  {
+    buffer<int> bIn1{in1, range{dataSize}};
+    buffer<int> bIn2{in2, range{dataSize}};
+    buffer<int> bIn3{in3, range{dataSize}};
+    buffer<int> bTmp{tmp, range{dataSize}};
+    buffer<int> bOut{out, range{dataSize}};
+
+    ext::codeplay::experimental::fusion_wrapper fw{q};
+    fw.start_fusion();
+
+    assert(fw.is_in_fusion_mode() && "Queue should be in fusion mode");
+
+    q.submit([&](handler &cgh) {
+      auto accIn1 = bIn1.get_access(cgh);
+      auto accIn2 = bIn2.get_access(cgh);
+      auto accTmp = bTmp.get_access(
+          cgh, sycl::ext::codeplay::experimental::property::promote_private{});
+      cgh.parallel_for<class KernelOne>(numVec, [=](id<1> i) {
+        size_t offset = i;
+        vec<int, VEC> in1;
+        in1.load(offset, accIn1.get_pointer());
+        vec<int, VEC> in2;
+        in2.load(offset, accIn2.get_pointer());
+        auto tmp = in1 + in2;
+        tmp.store(offset, accTmp.get_pointer());
```

Comment:
What does the spec say for a scenario where we use `multi_ptr` with a `__global` address space for `tmp`? Is it a user error? In either case, worth adding a test, I think?

Reply:
The internalization property specified by the user overrides the explicit address space and allows the JIT compiler to perform the internalization. The JIT compiler will still check that this does not result in any type errors and not perform the internalization if it would. 

I've added a test for a `multi_ptr` (`decorated_global_ptr` to be more specific).